### DISCUSSION TITLE
Problem about linking external library

### DISCUSSION THREAD

**Discussion by yuezhiying6** [idaholab/moose#18605] [Q&A General] [answered]
Dear all,
I am now linking a external FORTRAN library ti my code. I add the "ADDITIONAL_LIBS += -L/path/to/my/lib -lname " to my Makefile. However, when I use "make" command to compile my code, there will be an error occurs:

/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: warning: libgfortran.so.3, needed by /home/yzy/projects/full_couple/lib/libcobra.so, not found (try using -rpath or -rpath-link)
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_st_open@GFORTRAN_1.0'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_st_read@GFORTRAN_1.0'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_transfer_real_write@GFORTRAN_1.4'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_transfer_array@GFORTRAN_1.0'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_cpu_time_8@GFORTRAN_1.0'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_st_rewind@GFORTRAN_1.0'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_st_write_done@GFORTRAN_1.0'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_cpu_time_4@GFORTRAN_1.0'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_transfer_array_write@GFORTRAN_1.4'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_transfer_character_write@GFORTRAN_1.4'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_stop_numeric_f08@GFORTRAN_1.4'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_transfer_integer_write@GFORTRAN_1.4'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_transfer_character@GFORTRAN_1.0'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_stop_string@GFORTRAN_1.0'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `memcpy@GLIBC_2.14'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_transfer_real@GFORTRAN_1.0'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_st_write@GFORTRAN_1.0'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_transfer_integer@GFORTRAN_1.0'
/home/yzy/miniconda3/envs/moose/bin/../lib/gcc/x86_64-conda-linux-gnu/9.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/yzy/projects/full_couple/lib/libcobra.so: undefined reference to `_gfortran_st_read_done@GFORTRAN_1.0'
collect2: error: ld returned 1 exit status
/home/yzy/projects/moose/framework/app.mk:406: recipe for target '/home/yzy/projects/full_couple/full_couple-opt' failed
make: *** [/home/yzy/projects/full_couple/full_couple-opt] Error 1

I install my MOOSE through conda. Because I am not familiar with the ubuntu system, I don't know how to link this library successfully. 

Any advice would be greatly appreciated!
Zhiying

**Comment by GiudGiud** [comment, marked as answer]
Hello

It looks like your compiler is not finding libgfortran, needed to use compiled fortran code. 
Let's try to locate it first, either with
`whereis libgfortran` or `ldconfig -p | grep libgfortran`

If that does not work, you'll need to download it. 
@milljm does it come with the conda install?

Guillaume

**Reply by milljm** [reply]
Our latest environment installs libgortran 5:

```pre
(moose) [CentOS-8][~]> conda list | grep libgfortran
libgfortran-ng            9.4.0                h69a702a_8    conda-forge
libgfortran5              9.4.0                h62347ff_8    conda-forge
```

If your already-built-app was built using an older Fortran compiler, it will be necessary to either:

- rebuild that app using the same Fortran compiler MOOSE requires (libgfortran.so.5)
or
- remove the Conda MOOSE environment and install the necessary older Fortran compiler your app requires, and then build PETSc, libMesh, and MOOSE using this older Fortran compiler, so that everything matches.

**Reply by yuezhiying6** [reply]
Thanks for your reply very much, @milljm  and @GiudGiud. Your advice is of great help to me in solving the problem. I will try to recompile my code with libgfortran-5.
